### PR TITLE
fix: scope ClusterRole webhook and CRD permissions with resourceNames

### LINKED ISSUE #12836: Controller ClusterRole grants update on all Mutating/ValidatingWebhookConfigurations and CRDs cluster-wide
Controller ClusterRole grants update on all Mutating/ValidatingWebhookConfigurations and CRDs cluster-wide
  |  
-- | --
Severity | High
Location | upstream/kueue/src/config/components/rbac/role.yaml:54-71 (ClusterRole rule: admissionregistration.k8s.io {mutating,validating}webhookconfigurations verbs get,list,watch,update; apiextensions.k8s.io customresourcedefinitions verbs get,list,watch,update)
Location | upstream/kueue/src/pkg/util/cert/cert.go:40-42 (kubebuilder:rbac markers requesting cluster-wide webhookconfiguration update)
Location | upstream/kueue/src/pkg/visibility/server.go:77-79 (Duplicate kubebuilder:rbac markers)
Category | OWASP K02 Overly Permissive RBAC — Improper Privilege Management

Description

The kueue-controller-manager ClusterRole grants get/list/watch/update on ALL mutatingwebhookconfigurations and validatingwebhookconfigurations (and on all CRDs) without a resourceNames restriction. The legitimate need is CA-bundle injection into Kueue's own webhook configs and conversion-webhook CA injection into Kueue CRDs. An attacker who obtains the controller's service-account token (see FIND-001) or achieves code execution in the pod can rewrite any other operator's webhook clientConfig.url to an attacker endpoint, set failurePolicy: Ignore on security admission webhooks (Gatekeeper/Kyverno/PSA), or inject a malicious caBundle — defeating cluster-wide admission control. CIS Kubernetes Benchmark 5.1.1/5.1.3 (minimize wildcard/broad ClusterRole grants) and DISA STIG V-254800 apply.

config/components/rbac/role.yaml

```yaml
- apiGroups:
  - admissionregistration.k8s.io
  resources:
  - mutatingwebhookconfigurations
  - validatingwebhookconfigurations
  verbs:
  - get
  - list
  - update
  - watch
```
*Attack pattern*

Compromised kueue SA token → kubectl --token patch validatingwebhookconfiguration gatekeeper-validating-webhook-configuration --type=json -p '[{"op":"replace","path":"/webhooks/0/failurePolicy","value":"Ignore"}]' → policy bypass.

*Remediation*

Add resourceNames: [kueue-mutating-webhook-configuration, kueue-validating-webhook-configuration] to the admissionregistration rule and resourceNames listing only Kueue CRDs to the apiextensions rule. controller-gen does not emit resourceNames from markers, so maintain this as a kustomize patch (or move CA injection to cert-manager's cainjector and drop the verb entirely).

### PR DESCRIPTION
**Pull request by prash2512**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

The controller ClusterRole grants `get/list/watch/update` on **all** `MutatingWebhookConfigurations` `ValidatingWebhookConfigurations`, and `CustomResourceDefinitions` cluster-wide. The only legitimate need is CA-bundle injection into Kueue's own webhook configs and the 5 CRDs with conversion webhooks.
  
This PR adds `resourceNames` to the kubebuilder RBAC markers in `pkg/util/cert/cert.go`, scoping access to only Kueue's own resources. The issue's remediation suggested a kustomize patch since "controller-gen does not emit resourceNames from markers," but this was fixed in [controller-tools#413](https://github.com/kubernetes-sigs/controller-tools/pull/413) (v0.2.8) - Kueue uses v0.20.1, so the markers work directly.

The webhook configuration names are install-dependent, so the markers carry the **unprefixed** name and each installer applies its own prefix: Helm via `kueue.fullname` (the same expression `webhook/manifests.yaml` uses), kustomize via `namePrefix` plus a `nameReference` declaration. A non-default kustomize `namePrefix` works as a result. `list`/`watch` stay unscoped — RBAC cannot restrict them by name and the cert rotator's informers need them.

Also grants `get` on `clusterprofiles.multicluster.x-k8s.io`: `pkg/config` reads that CRD to detect whether ClusterProfile support is installed, and scoping CRD access had revoked it.

#### Which issue(s) this PR fixes:

Fixes #12836

#### Special notes for your reviewer:

Verified locally that `make manifests` generates the expected `resourceNames` in `config/components/rbac/role.yaml`. No code logic changes — only static RBAC manifest and Helm chart updates.

#### Does this PR introduce a user-facing change?

```release-note
Fixed overly broad ClusterRole permissions by scoping webhook configuration and CRD access to only Kueue's own resources using `resourceNames`
```

## Summary by CodeRabbit

* **Security**
  * Further tightened RBAC for admission webhooks by scoping access to specific mutating/validating webhook configurations, limiting `get`/`update` accordingly.
  * Kept `list`/`watch` broad for webhook configurations while restricting name-sensitive access.
  * Reduced CRD permissions by limiting `get`/`update` to only the required Kueue custom resources (plus `clusterprofiles`).
  * Added kustomize wiring to keep webhook-related `resourceNames` aligned after name changes.
* **Tests**
  * Added RBAC tests to validate webhook and CRD `resourceNames` scoping behavior.

### CODE REVIEW

File: pkg/util/cert/cert.go [hunk 1]
```diff
@@ -37,9 +37,9 @@ const (
 	webhookServiceSuffix = "-webhook-service"
 )
 
-// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=mutatingwebhookconfigurations,verbs=get;list;watch;update
-// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=validatingwebhookconfigurations,verbs=get;list;watch;update
-// +kubebuilder:rbac:groups="apiextensions.k8s.io",resources=customresourcedefinitions,verbs=get;list;watch;update
+// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=mutatingwebhookconfigurations,resourceNames=kueue-mutating-webhook-configuration,verbs=get;list;watch;update
+// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=validatingwebhookconfigurations,resourceNames=kueue-validating-webhook-configuration,verbs=get;list;watch;update
+// +kubebuilder:rbac:groups="apiextensions.k8s.io",resources=customresourcedefinitions,resourceNames=clusterqueues.kueue.x-k8s.io;cohorts.kueue.x-k8s.io;localqueues.kueue.x-k8s.io;multikueueclusters.kueue.x-k8s.io;workloads.kueue.x-k8s.io,verbs=get;list;watch;update
```
[the remark this answers is not in the document]
Comment (prash2512) [reply pkg/util/cert/cert.go]: Template resourceNames for webhook configurations to match the deployed resource names.

The webhook configurations in the chart manifests are deployed using templated names ({{ include "kueue.fullname" . }}-mutating-webhook-configuration and {{ include "kueue.fullname" . }}-validating-webhook-configuration). However, this generated RBAC role hardcodes them with the kueue- prefix. If a user installs the chart with a non-default release name, the cert-controller will receive "Forbidden" errors because the deployed object names will not match this allowlist.

Since this file is generated by yaml-processor, please ensure the generator script or configuration is updated to substitute the hardcoded names with Helm template directives. As per path instructions, we must verify Helm template correctness.

**Comment by mimowo** [comment]
cc @kannon92 ptal. Is this something we would like to cherry-pick?

File: pkg/util/cert/cert.go [hunk 2]
```diff
@@ -37,9 +37,12 @@ const (
 	webhookServiceSuffix = "-webhook-service"
 )
 
-// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=mutatingwebhookconfigurations,verbs=get;list;watch;update
-// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=validatingwebhookconfigurations,verbs=get;list;watch;update
-// +kubebuilder:rbac:groups="apiextensions.k8s.io",resources=customresourcedefinitions,verbs=get;list;watch;update
+// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=mutatingwebhookconfigurations,verbs=list;watch
```
Comment (kannon92) [inline pkg/util/cert/cert.go:46]: Can you only add the resourceName fix here?

I see some extra cleanup (removing some verbs). We maybe could use this but that should be a separate PR.

**Comment by kannon92** [comment]
> cc @kannon92 ptal. Is this something we would like to cherry-pick?

Yea.

File: pkg/util/cert/cert.go (hunk 2 above)
Comment (prash2512) [reply pkg/util/cert/cert.go:46]: Hi @kannon92 

Thanks for the review!

The verb split is necessary - resourceNames only matches requests that target a specific named resource. list/watch requests don't include a resource name, so they can never match a rule with resourceNames set. Since RBAC is default-deny, that would block the cert-controller's informers from starting. The two rules keep list/watch unscoped and scope only get/update with resourceNames.

**Comment by prash2512** [comment]
Hi @kannon92 

I have replied to your comment
Can I get a follow up review?
Thanks

File: pkg/util/cert/cert.go [hunk 3]
```diff
@@ -37,9 +37,12 @@ const (
...
+// +kubebuilder:rbac:groups="admissionregistration.k8s.io",resources=mutatingwebhookconfigurations,resourceNames=kueue-mutating-webhook-configuration,verbs=get;update
```
Comment (kannon92) [inline pkg/util/cert/cert.go]: AI review:

We should revert any fix that assumes hard coding of kueue-mutating-webhook-configuration or kueue-validating-webhook-configuration.


 Backwards-compatibility concern (real, will break deployments): hardcoded webhook resourceNames vs. templated/derived object names                                                                                                                                                       
                                                                                                                                                                                                                                                                                          
 - The RBAC change hardcodes:                                                                                                                                                                                                                                                             
     - resourceNames: [kueue-mutating-webhook-configuration]                                                                                                                                                                                                                              
     - resourceNames: [kueue-validating-webhook-configuration]                                                                                                                                                                                                                            
 - But pkg/util/cert/cert.go derives the actual webhook configuration names dynamically at runtime from cfg.InternalCertManagement.WebhookServiceName via deriveWebhookBaseName + buildWebhookConfigurationName — and the Helm chart's own webhook manifests                              
   (charts/kueue/templates/webhook/manifests.yaml) name these objects {{ include "kueue.fullname" . }}-mutating-webhook-configuration, i.e. they change with nameOverride/fullnameOverride/release name.                                                                                  
 - pkg/util/cert/cert_test.go explicitly tests non-default names like test-kueue, prod-kueue, my-company-staging-kueue — proving this is a supported customization, not an edge case.                                                                                                     
 - Result: any user who installs the Helm chart with a custom release/full name (a common practice, e.g. multiple Kueue installs per cluster) will get a manager whose ClusterRole is scoped to the default webhook config names, while the actual objects created have a different name. 
   get/update on the real MutatingWebhookConfiguration/ValidatingWebhookConfiguration will then be denied by RBAC, breaking CA-bundle injection and cert rotation — a functional regression, not just a permissions tightening, for existing customized deployments.                      
 - The kustomize-based install (config/) happens to line up today because namePrefix: kueue- matches the hardcoded default, but that's coincidental and fragile (breaks if namePrefix is ever changed).                                                                                   
                                                                                                                                                                                                                                                                                          
 CRD resourceNames are fine: the 5 hardcoded CRD names (clusterqueues, cohorts, localqueues, multikueueclusters, workloads.kueue.x-k8s.io) are fixed API group/resource names, not affected by release naming, and match the CRDConversion list in buildCertRotatorConfig. No concern     
 there — as long as no future CRD needing conversion-webhook CA injection is added without updating this list in lockstep (should flag in review checklist, but not a backwards-compat break today).                                                                                      
                                                                                                                                                                                                                                                                                          
 Recommendation: Don't hardcode webhook resourceNames as literals. Either:                                                                                                                                                                                                                
 1. Template the resourceNames in the Helm chart using {{ include "kueue.fullname" . }} (same pattern as manifests.yaml), and similarly parameterize config/components/rbac/role.yaml if it's meant to support prefix overrides, or                                                       
 2. Drop the resourceNames restriction on the webhook configurations specifically (keep it for CRDs, which are truly fixed names) since scoping something whose name is only known at runtime/deploy-time via templating is exactly the case resourceNames handles poorly for reusable    
    charts.

Comment (prash2512) [reply pkg/util/cert/cert.go]: Updated 

Markers now carry the unprefixed name and each installer adds its own - Helm via `kueue.fullname`, kustomize via `namePrefix` plus a `nameReference` declaration. Verified across release names `kueue`, `foo-kueue`, `kueue-populator`; chart unit test added.

Kept `resourceNames` rather than dropping them, since that grant is what #12836 is about.

**Comment by prash2512** [comment]
Hi @kannon92

Thanks for the review. I have addressed the feedback

Adding `resourceNames` to the webhook markers had two side effects:

1. The hardcoded `kueue-` only works if the install is named `kueue`. Under any other Helm release name the manager can't update its own webhook configs, so it never starts. The markers now carry the unprefixed name and each installer adds its own prefix. Helm gets it from `kueue.fullname`, kustomize from `namePrefix`. Custom `namePrefix` works now too.

2. Scoping CRD `get` broke `pkg/config`'s read of `clusterprofiles.multicluster.x-k8s.io`. Forbidden isn't NotFound, so it never hits the "CRD not installed" path and the manager just exits when `MultiKueueClusterProfile` is on. Added back as a `get`-only rule.

On (2), I put the marker in `pkg/config/config.go` next to the call that needs it, which meant adding `./pkg/config/...` to controller-gen's paths. The other option was `cert.go`, since the existing CRD markers live there, but the permission has nothing to do with cert rotation. Any preference?

Two follow-ups I'd like your view on:
1.  Instead of doing a `get` on the CRD object, `pkg/config` could ask the API server whether `clusterprofiles` is served under `multicluster.x-k8s.io/v1alpha1`. Any authenticated client can query discovery, so that needs no RBAC and the `get` rule could go away completely. That changes how MultiKueue detects the feature at startup rather than anything about RBAC, so it felt out of place in this PR. I can file a separate issue, if its worth fixing?
2. No Helm e2e job installs under a non-default release name, so (1) wasn't covered there. A chart unit test covers it now. Doing it properly in e2e means touching shared helpers that every suite uses. Can follow up with a separate issue if we need an e2e test exercising this path?

**Review by kannon92 — COMMENTED** [review]
/lgtm

File: pkg/config/config.go [hunk 4]
```diff
@@ -235,6 +235,16 @@ func Load(scheme *runtime.Scheme, configFile string) (ctrl.Options, configapi.Co
...
 }
 
+// The rule below covers the ClusterProfile CRD object itself, not ClusterProfiles. Without
+// it a cluster running with MultiKueueClusterProfile enabled gets Forbidden rather than
+// NotFound, which is not the "CRD absent" case configureClusterProfileCacheWithClient
+// tolerates, so the manager exits during startup.
+//
+// TODO(#TBD): this is a capability probe rather than a data read, so it belongs on the
+// discovery API, which every authenticated client may read. Switching to discovery lets
+// this rule be dropped entirely.
```
Comment (mimowo) [inline pkg/config/config.go]: Either drop the T`ODO(#TBD`) or open a dedicated issue and refrerence it here as `TODO(#XYZ)`

**Comment by mimowo** [comment]
LGTM overall, I'm concidering a cherrypick since this is considered a bugfix, any concerns @kannon92 @tenzen-y ?

**Comment by kannon92** [comment]
> LGTM overall, I'm concidering a cherrypick since this is considered a bugfix, any concerns @kannon92 @tenzen-y ?

No concerns from me. This came up during security review so its probably worth a carry.

File: pkg/config/config.go (hunk 4 above)
Comment (prash2512) [reply pkg/config/config.go]: @mimowo Thanks for the review
Dropped the `TODO(#TBD)`. I looked into moving that check to API discovery and it isn't worth it. CRD get seems to be a better solution without pinning the API version.

**Comment by mimowo** [comment]
/lgtm
/approve
/cherrypick release-0.19
/cherrypick release-0.18
Thank you for this security hardening fix 👍
